Add fully-qualified class names of jar verification providers

KostasTsiounis opened the pull request:
Additional entries, pertaining to the fully-qualified class names of the providers, are added to the list utilized to populate the `ProviderList` that is in turn used for jar verification.

This PR is related to https://github.com/eclipse-openj9/openj9/issues/19825.

Signed-off-by: Kostas Tsiounis <kostas.tsiounis@ibm.com>

keithc-ca commented on src/java.base/share/classes/sun/security/jca/Providers.java:
```diff
@@ -90,14 +90,18 @@ private Providers() {
-        "SUN",
-        "SunRsaSign",
+        "SUN", "sun.security.provider.Sun",
+        "SunRsaSign", "sun.security.rsa.SunRsaSign",
         // Note: when SunEC is in a signed JAR file, it's not signed
         // by EC algorithms. So it's still safe to be listed here.
-        "SunEC",
-        "SunJCE",
+        "SunEC", "sun.security.ec.SunEC",
+        "SunJCE", "com.sun.crypto.provider.SunJCE",
     };
```
To avoid potentially loading providers twice, perhaps we should check `RestrictedSecurity.isEnabled()` to decide whether to use a list of fully-qualified names instead of the normal list?

KostasTsiounis replied:
Changed to have two separate arrays and a check to decide which one to use.

keithc-ca commented on src/java.base/share/classes/sun/security/jca/Providers.java:
```diff
@@ -100,6 +100,20 @@ private Providers() {
         "SunJCE",
     };
 
+    // Hardcoded fully qualified names of providers to use for JAR
```
This should be hyphenated: "fully-qualified class names".

KostasTsiounis replied:
Changed.

keithc-ca commented on src/java.base/share/classes/sun/security/jca/Providers.java:
```diff
@@ -100,6 +100,20 @@ private Providers() {
     };
 
+    // Hardcoded fully qualified names of providers to use for JAR
+    // verification when RestrictedSecurity is enabled.
+    // (Similar to jarVerificationProviders array)
```
Missing period, suggest:
```
    // verification when RestrictedSecurity is enabled (similar to
    // jarVerificationProviders array).
```

KostasTsiounis replied:
Changed to suggestion.

keithc-ca commented on src/java.base/share/classes/sun/security/jca/Providers.java:
```diff
@@ -115,7 +129,13 @@ public static Provider getSunProvider() {
     public static Object startJarVerification() {
         ProviderList currentList = getProviderList();
-        ProviderList jarList = currentList.getJarList(jarVerificationProviders);
+        ProviderList jarList;
+        if (RestrictedSecurity.isEnabled()) {
+            jarList = currentList.getJarList(jarVerificationProvidersRS);
+        } else {
+            jarList = currentList.getJarList(jarVerificationProviders);
+        }
+
```
Suggest either capturing the argument to `getJarList()` in a local, or simply making it conditional
```java
        ProviderList jarList = currentList.getJarList(
                RestrictedSecurity.isEnabled()
                        ? restrictedJarVerificationProviders
                        : jarVerificationProviders);
```
We don't need blank line 138.

KostasTsiounis replied:
Changed to suggestion.

keithc-ca commented on src/java.base/share/classes/sun/security/jca/Providers.java:
```diff
@@ -100,6 +100,20 @@ private Providers() {
+    // (Similar to jarVerificationProviders array)
+    //
+    // MUST NOT be on the bootclasspath and not in signed JAR files.
+    private static final String[] jarVerificationProvidersRS = {
```
Suggest a better name is `restrictedJarVerificationProviders` (rather than abbreviate in `RS`).

KostasTsiounis replied:
Changed to suggestion.

keithc-ca commented:
The message in the first commit is missing "is" in "that _is_ utilized for Jar verification".

KostasTsiounis commented:
> The message in the first commit is missing "is" in "that _is_ utilized for Jar verification".

I fixed the message of the first commit and squashed.

keithc-ca commented on src/java.base/share/classes/sun/security/jca/Providers.java:
```diff
@@ -100,6 +100,20 @@ private Providers() {
         "SunJCE",
     };
 
+    // Hardcoded fully-qualified names of providers to use for JAR
```
These are "class" names: The comment should say so.

KostasTsiounis replied:
Updated to indicate that.

keithc-ca commented:
Jenkins test sanity zlinux jdknext
